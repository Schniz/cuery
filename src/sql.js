// @flow

import Query from './Query';
import buildQueryString from "./buildQueryString";
import { type SqlFunctionParameter } from './types';

export default function sql(
  strings: Array<string>,
  ...functionsOrParameters: Array<SqlFunctionParameter | Query>
): Query {
  const query = buildQueryString(strings, functionsOrParameters);

  const params = functionsOrParameters.reduce((acc, curr) => {
    if (curr instanceof Query) {
      acc.push(...curr.queryMeta.params);
    } else {
      acc.push(curr);
    }
    return acc;
  }, []);

  return new Query({ query, params });
}
