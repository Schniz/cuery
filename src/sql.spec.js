import sql from "./sql";

it("it builds a query object", () => {
  const result = sql`select * from users where id = ${'some id'}`;
  expect(result).toMatchSnapshot();
});

it("composes", () => {
  const first = sql`select * from users where id = ${'first query param'}`;
  const withName = sql`select * from (${first}) where name = ${'second query param'}`;
  const onlyFirstWithName = sql`select * from (${withName}) limit 1`;
  expect(onlyFirstWithName).toMatchSnapshot();
})
