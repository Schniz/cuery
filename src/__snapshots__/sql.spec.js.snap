// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`composes 1`] = `
Query {
  "queryMeta": Object {
    "params": Array [
      "first query param",
      "second query param",
    ],
    "query": "select * from (select * from (select * from users where id = $1) where name = $2) limit 1",
  },
}
`;

exports[`it builds a query object 1`] = `
Query {
  "queryMeta": Object {
    "params": Array [
      "some id",
    ],
    "query": "select * from users where id = $1",
  },
}
`;
